flight_computer/lib/opt3001.py

```diff
@@ -0,0 +1,156 @@
+import time
+
+from adafruit_bus_device.i2c_device import I2CDevice
+
+# Can be one of four possible values
+# This assumes we connect the ADDR pin to GND
+# See 8.3.4.1 in the specification
+DEVICE_DEFAULT_I2C_ADDR = 0x44
+
+# MAIN REGISTERS FOR PROGRAMMING DEVICE
+RESULT = 0x00
+CONFIGURATION = 0x1
+
+# CONFIGURATION SETTINGS
+AUTO_RANGE_SETTING = 0b1100
+CONTINUOUS_CONVERSIONS = 0b11
+DEFAULT_CONVERSION_TIME = 800
+MAX_CONVERSION_TIME = 800
+
+# FOR INTERRUPT MECHANISMS, UNUSED
+LOW_LIMIT = 0x2
+HIGH_LIMIT = 0x3
+
+# CAN BE READ TO GET ID
+MANUFACTURER_ID = 0x7E
+DEVICE_ID = 0x7F
+
+REGISTER_SIZE = 2
+
+
+# EXTRACT k bits from and including pos p
+# using n ... 0 indexing
+def extract_bits(bits, pos, k):
+    # first chop off everything exclusively after pos
+    # then get k inclusive least significant bits from the 0 index
+    return (bits >> pos) & ((1 << k) - 1)
+
+
+class Opt3001:
+    """OPT3001 Sun Sensor Driver"""
+
+    def __init__(self, i2c_bus, address=DEVICE_DEFAULT_I2C_ADDR, conversion_time=DEFAULT_CONVERSION_TIME):
+        # Initialize and Configure the sun sensor
+
+        # Initialize i2c device from i2c passed in
+        self.i2c_device = I2CDevice(i2c_bus, address)
+
+        # Initialize a fixed buffer to read and write from
+        self.buf = bytearray(3)
+
+        # DEVICE ID SHOULD BE 0x3001
+        self.read_value(DEVICE_ID)
+        if self.buf != bytes([DEVICE_ID, 0x30, 0x01]):
+            raise Exception("ERROR: Could not read correct device ID from bus provided")
+
+        # MANUFACTURER ID Should be 0x5449
+        self.read_value(MANUFACTURER_ID)
+        if self.buf != bytes([MANUFACTURER_ID, 0x54, 0x49]):
+            raise Exception("ERROR: Could not read correct manufacturer ID from bus provided")
+
+        self.configure(conversion_time)
+
+    # DeInitialize the sun sensor
+    def __del__(self):
+        # put sensor in low power mode since we are done using it
+        # specifically, low power is M[1:0] set to 0b00,
+        # but we can just reset everything
+        self.buf[1] = 0x00
```

Comment:
I'm not sure if setting the whole configuration register to 0 Is the best option. Maybe we should read the current value and then only write the mode setting?